src/Microsoft.Data/DataFrame.IO.cs

```diff
@@ -0,0 +1,261 @@
+            Type res = typeof(string);
+            int nbline = 0;
+            foreach (var line in read)
+            {
+                if (col >= line.Length)
+                    throw new FormatException(string.Format(Strings.LessColumnsThatExpected, nbline + 1)); // Will this really work?
+
+                var val = line[col];
+                try
+                {
+                    bool.Parse(val);
+                    res = DetermineType(nbline == 0, typeof(bool), res);
+                    ++nbline;
+                    continue;
+                }
+                catch (Exception)
+                {
+                    if (string.IsNullOrEmpty(val))
+                    {
+                        res = DetermineType(nbline == 0, typeof(bool), res);
+                        continue;
+                    }
+                }
+                try
+                {
+                    float.Parse(val);
+                    res = DetermineType(nbline == 0, typeof(float), res);
+                    ++nbline;
+                    continue;
+                }
+                catch (Exception)
+                {
+                    if (string.IsNullOrEmpty(val))
+                    {
+                        res = DetermineType(nbline == 0, typeof(float), res);
+                        continue;
+                    }
+                }
+                res = DetermineType(nbline == 0, typeof(string), res);
+                ++nbline;
+            }
+            return res;
+        }
+
+        private static Type DetermineType(bool first, Type suggested, Type previous)
+        {
+            if (first)
+                return suggested;
+            else
+                return MaxKind(suggested, previous);
+        }
+
+        private static Type MaxKind(Type a, Type b)
+        {
+            if (a == typeof(string) || b == typeof(string))
+                return typeof(string);
+            if (a == typeof(float) || b == typeof(float))
+                return typeof(float);
+            if (a == typeof(bool) || b == typeof(bool))
+                return typeof(bool);
+            return typeof(string);
+        }
+
+        /// <summary>
+        /// Reads a text file as a IDataView.
+        /// Follows pandas API.
+        /// </summary>
+        /// <param name="filename">filename</param>
+        /// <param name="sep">column separator</param>
+        /// <param name="header">has a header or not</param>
+        /// <param name="names">column names (can be empty)</param>
+        /// <param name="dtypes">column types (can be empty)</param>
+        /// <param name="nrows">number of rows to read</param>
+        /// <param name="guess_rows">number of rows used to guess types</param>
+        /// <param name="encoding">text encoding</param>
+        /// <param name="index">add one column with the row index</param>
+        /// <returns>DataFrame</returns>
+        public static DataFrame ReadCsv(string filename,
```

Comment:
I almost wonder if these methods should be external to the `DataFrame`. That way we can ensure anyone can take any data (like from SQL, or elsewhere) and materialize a DataFrame.

Reply:
By external, do you mean an API such as 

`ReadCsv(filename, out DataFrame dataFrame)`?

Reply:
I mean not on the `DataFrame` class. The reasoning is that other code outside of the current assembly should be able to load `DataFrame` objects from different sources - SQL, Parquet, in-memory objects, etc. If `ReadCsv` is on the `DataFrame` class itself, it has access to `private` fields/methods that other code won't be able to use. This means it gets `special` treatment. If someone else wanted to write a "create a DataFrame from some data source" method, would they be able to do it?